sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -717,47 +717,121 @@ gnrc_netif_t *gnrc_netif_get_by_prefix(const ipv6_addr_t *prefix)
     return best_netif;
 }
 
+static int _netif_ops_set_helper(gnrc_netif_t *netif, netopt_t opt,
+                                 void *data, uint16_t data_len)
+{
+    gnrc_netapi_opt_t netapi_opt = {
+        .opt = opt,
+        .data = data,
+        .data_len = data_len,
+    };
+    return netif->ops->set(netif, &netapi_opt);
+}
+
 int gnrc_netif_ipv6_group_join_internal(gnrc_netif_t *netif,
                                         const ipv6_addr_t *addr)
 {
-    unsigned idx = UINT_MAX;
+    uint8_t group_data[GNRC_NETIF_L2ADDR_MAXLEN];
+    int res;
 
+    /* can be called out of lock */
+    res = gnrc_netif_ipv6_group_to_l2_group(netif, addr, group_data);
     gnrc_netif_acquire(netif);
+    if (res > 0) {
+
+        res = _netif_ops_set_helper(netif, NETOPT_L2_GROUP,
+                                    group_data, (uint16_t)res);
+        /* link layer does not support multicast, but we can still use
+         * broadcast */
+        if ((res != -ENOTSUP) && (res < 0)) {
+            goto end;
+        }
+    }
+    /* link layer does not support multicast, but we can still use
+     * broadcast */
+    else if (res != -ENOTSUP) {
+        goto end;
+    }
+    res = -ENOMEM;
     for (unsigned i = 0; i < GNRC_NETIF_IPV6_GROUPS_NUMOF; i++) {
         if (ipv6_addr_equal(&netif->ipv6.groups[i], addr)) {
             gnrc_netif_release(netif);
             return i;
         }
-        if ((idx == UINT_MAX) &&
+        if ((res < 0) &&
             (ipv6_addr_is_unspecified(&netif->ipv6.groups[i]))) {
-            idx = i;
+            res = i;
         }
     }
-    if (idx == UINT_MAX) {
-        gnrc_netif_release(netif);
-        return -ENOMEM;
+    if (res < 0) {
+        goto end;
     }
-    memcpy(&netif->ipv6.groups[idx], addr, sizeof(netif->ipv6.groups[idx]));
+    memcpy(&netif->ipv6.groups[res], addr, sizeof(netif->ipv6.groups[res]));
     /* TODO:
      *  - MLD action
      */
+end:
     gnrc_netif_release(netif);
-    return idx;
+    return res;
```

Comment:
I think this whole function would be much more readable when not using `goto` to save one line of code and avoid reusing `res` for completely different things. E.g. at some point `res` was the size of the group, at other points it was a error code to return, and at one more point it could be an index of an IPv6 address.